src/MessageFormatter.php

```diff
@@ -0,0 +1,180 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Log;
+
+use DateTime;
+
+use function gettype;
+use function implode;
+use function is_array;
+use function is_string;
+use function microtime;
+use function sprintf;
+use function strpos;
+
+final class MessageFormatter
+{
+    /**
+     * @var callable|null PHP callable that returns a string representation of the log message.
+     *
+     * If not set, {@see MessageFormatter::defaultFormat()} will be used.
+     *
+     * The signature of the callable should be `function (array $message): string;`.
+     */
+    private $format;
+
+    /**
+     * @var callable|null PHP callable that returns a string to be prefixed to every exported message.
+     *
+     * If not set, {@see MessageFormatter::getPrefix()} will be used, which prefixes
+     * the message with context information such as user IP, user ID and session ID.
+     *
+     * The signature of the callable should be `function (array $message): string;`.
+     */
+    private $prefix;
+
+    /**
+     * @var string The date format for the log timestamp. Defaults to `Y-m-d H:i:s.u`.
+     */
+    private string $timestampFormat = 'Y-m-d H:i:s.u';
+
+    /**
+     * Sets the format for the string representation of the log message.
+     *
+     * @param callable $format The PHP callable to get a string representation of the log message.
+     *
+     * @see MessageFormatter::$format
+     */
+    public function setFormat(callable $format): void
+    {
+        $this->format = $format;
+    }
+
+    /**
+     * Sets a PHP callable that returns a string to be prefixed to every exported message.
+     *
+     * @param callable $prefix The PHP callable to get a string prefix of the log message.
+     *
+     * @see MessageFormatter::$prefix
+     */
+    public function setPrefix(callable $prefix): void
+    {
+        $this->prefix = $prefix;
+    }
+
+    /**
+     * Sets a date format for the log timestamp.
+     *
+     * @param string $timestampFormat The date format for the log timestamp.
+     *
+     * @see MessageFormatter::$timestampFormat
+     */
+    public function setTimestampFormat(string $timestampFormat): void
+    {
+        $this->timestampFormat = $timestampFormat;
+    }
+
+    /**
+     * Formats a log message for display as a string.
+     *
+     * The message structure follows that in {@see MessageCollection::$messages}.
+     *
+     * @param array $message The log message to be formatted.
+     *
+     * @throws LogRuntimeException for a callable "format" that does not return a string.
+     *
+     * @return string The formatted log message.
+     */
+    public function format(array $message): string
+    {
+        if ($this->format === null) {
+            return $this->defaultFormat($message);
+        }
+
+        $formatted = ($this->format)($message);
+
+        if (!is_string($formatted)) {
+            throw new LogRuntimeException(sprintf(
```

Comment:
We can use `\RuntimeException` here since noone will ever catch it.

Reply:
Let's leave `LogRuntimeException`? This exception is thrown by the same condition in other classes of this package.